Core/include/Acts/Clusterization/Clusterization.ipp

```diff
@@ -110,60 +118,69 @@ struct Connections<2> : public ConnectionsBase<4> {
 
 // Cell collection logic
 template <typename Cell, typename Connect, std::size_t GridDim>
-Connections<GridDim> getConnections(typename std::vector<Cell>::iterator it,
-                                    std::vector<Cell>& set, Connect connect) {
+Connections<GridDim> getConnections(std::size_t idx, std::vector<Cell>& cells,
+                                    std::vector<Label>& labels,
+                                    Connect&& connect) {
   Connections<GridDim> seen;
-  typename std::vector<Cell>::iterator it_2{it};
 
-  while (it_2 != set.begin()) {
-    it_2 = std::prev(it_2);
+  for (std::size_t i = 0; i < idx; ++i) {
+    std::size_t idx2 = idx - i - 1;
+    ConnectResult cr = connect(cells[idx], cells[idx_2]);
 
-    ConnectResult cr = connect(*it, *it_2);
     if (cr == ConnectResult::eNoConnStop) {
       break;
     }
     if (cr == ConnectResult::eNoConn) {
       continue;
     }
     if (cr == ConnectResult::eConn) {
-      seen.buf[seen.nconn] = getCellLabel(*it_2);
+      seen.buf[seen.nconn] = labels[idx_2];
       seen.nconn += 1;
       if (seen.nconn == seen.buf.size()) {
         break;
       }
     }
   }
+
   return seen;
 }
 
 template <typename CellCollection, typename ClusterCollection>
-  requires(
-      Acts::Ccl::HasRetrievableLabelInfo<typename CellCollection::value_type> &&
-      Acts::Ccl::CanAcceptCell<typename CellCollection::value_type,
-                               typename ClusterCollection::value_type>)
-ClusterCollection mergeClustersImpl(CellCollection& cells) {
+  requires(Acts::Ccl::CanAcceptCell<typename CellCollection::value_type,
+                                    typename ClusterCollection::value_type>)
+ClusterCollection mergeClustersImpl(CellCollection& cells,
+                                    const std::vector<Label>& cellLabels,
+                                    const std::vector<std::size_t>& nClusters) {
   using Cluster = typename ClusterCollection::value_type;
 
-  if (cells.empty()) {
-    return {};
+  // Accumulate clusters into the output collection
+  ClusterCollection outv(nClusters.size());
```

Comment:
- name could be a bit improved
- do we need the default constructed elements? otherwise `reserve`+`push_back` might be a valid alternative